### PR TITLE
Refactoring Layout: Adding column widths, forms, and more.

### PR DESCRIPTION
**Pull request by aliabid94**
Lots of changes related to layout, sorry for the large PR. Fixes: #1162 

# Column width manipulation

Can now set width of columns using `scale` (relative widths) and `min_width`

# Replacing mobile collapse for flex based wrapping

Collapsing previously had issues because collapsing was defined by screen width, not element width. For example, a nested row would not collapse even if it had limited space, simply because the screen was large. Neither would a row in an embedded space. Now this works  

# Simplifying "form" visual generation

Combing adjacent form elements into a unified form visual was complex and a bit hacky. Previously there was a lot of logic to calculate which components were "first", "middle" or "last" and render their borders appropriately. This had a lot of css and js code, and had bugs (e.g.  if a "first" component was hidden, the second component had a missing border)

Simplified significantly. Now in the backend, adjacent form components are automatically combined into a gr.Form, and in the frontend we simply render a Form layout, which inherits its flex-direction.

# Automatic parent generation (for Forms and Tab)

To do the previous change (combining adjacent form components to be within a Form layout), I had to implement automatic parent generation. That means 
```python
with gr.Row():
  t1 = gr.Textbox()
  n1 = gr.Number()
```
automatically was converted to:
```python
with gr.Row():
  with gr.Form():
    t1 = gr.Textbox()
    n1 = gr.Number()
```

it was a simple change to extend this to tabs, so now:
```python
with gr.Tab("item1"):
  t1 = gr.Textbox()
with gr.Tab("item2"):
  t2 = gr.Textbox()
```
automatically adds the parent `Tabs` if missing, and is converted to:
```python
with gr.Tabs():
  with gr.Tab("item1"):
    t1 = gr.Textbox()
  with gr.Tab("item2"):
    t2 = gr.Textbox()
```

saving a lot of nesting. Note I also renamed `TabItem` to `Tab`, and have updated documentation. 

I have NOT implemented this for columns yet, but we could also apply this to columns, so
```python
with gr.Column():
  t1 = gr.Textbox()
with gr.Column():
  t2 = gr.Textbox()
```
automatically generates
```python
with gr.Row():
  with gr.Column():
    t1 = gr.Textbox()
  with gr.Column():
    t2 = gr.Textbox()
```

## Other

- Added "component-" prefix to autogenerated element ids since previous ids were only numbers, which is an invalid attribute and could not be used with CSS.

### CODE REVIEW

**Comment by abidlabs** [comment]
Thanks for fixing this issue @aliabid94! What about a slight modification of this API so that for width, you can accept a _ratio_ or a _factor_, similar to how flex works. So for example, if two elements are in a row, and the first has width=3, and the second has width=1, the first one will take 75% of the available width, while the second will take 25% of the available width (by default, elements could have a width factor of 1)? Tbh I'm not sure if this is a better approach, but I wanted to bring it up as the approach proposed in this PR could lead to weird conflicts (for example, if the percentages add up to more than 100%, etc.)

**Comment by abidlabs** [comment]
Wow this is good stuff @aliabid94!

**Review by pngwn — DISMISSED** [review]
Looks good other than one point:

> Note I also renamed TabItem to Tab, and have updated documentation.

This is a pretty major breaking change and will need to be reverted unless we want to release  4.0. Alternatively, we can deprecate `TabItem` and add `Tab` in addition to `TabItem` rather than replacing it.

**Comment by aliabid94** [comment]
Sorry should have been more clear, Tab has been added in addition to TabItem, which has been removed form the documentation but preserved for backwards compatibility.

File: gradio/blocks.py [hunk 1]
```diff
@@ -425,6 +425,7 @@ def get_block_instance(id: int) -> Block:
             block_config["props"].pop("type", None)
             block_config["props"].pop("name", None)
             style = block_config["props"].pop("style", None)
+            print("!!!", block_config)
```
Comment (abidlabs) [inline gradio/blocks.py:428]: Remove
```suggestion
            print("!!!", block_config)
```

File: gradio/blocks.py [hunk 2]
```diff
@@ -693,6 +694,35 @@ def get_config_file(self):
...
+                if not expected_parent or isinstance(block, expected_parent):
+                    pseudo_parent = None
+                    children.append(child)
+                else:
+                    if pseudo_parent is not None and isinstance(
+                        pseudo_parent, expected_parent
+                    ):
+                        pseudo_parent.children.append(child)
+                    else:
+                        pseudo_parent = expected_parent(render=False)
```
Comment (abidlabs) [inline gradio/blocks.py:714]: So `render()` is useful after all ;)

File: gradio/layouts.py [hunk 3]
```diff
@@ -55,49 +47,65 @@ def update(
...
-
+            warnings.warn("mobile_collapse is no longer supported.", DeprecationWarning)
         return self
 
 
 @document()
 class Column(BlockContext):
     """
-    Column is a layout element within Blocks that renders all children vertically.
+    Column is a layout element within Blocks that renders all children vertically. The widths of columns can be set through its arguments.
```
Comment (abidlabs) [inline gradio/layouts.py:68]:
```suggestion
    Column is a layout element within Blocks that renders all children vertically. The widths of columns can be set through the `scale` and `min_width` parameters.
```

**Comment by abidlabs** [comment]
LGTM! Some nits above

**Review by abidlabs — APPROVED** [review]

File: demo/blocks_plug/run.py [hunk 4]
```diff
@@ -22,13 +22,12 @@ def change_tab():
...
 with gr.Blocks() as demo:
     gr.Markdown("Three demos in one!")
-    with gr.Tabs(selected=1) as tabs:
-        with gr.TabItem("Text Identity", id=0):
-            identity_demo.render()
-        with gr.TabItem("Text Input", id=1):
-            input_demo.render()
-        with gr.TabItem("Text Static", id=2):
-            output_demo.render()
+    with gr.Tab("Text Identity", id=0):
```
Comment (freddyaboulton) [inline demo/blocks_plug/run.py:25]: This breaks the demo since the `tabs` variable is no longer defined!

File: demo/blocks_plug/run.py [hunk 5]
```diff
@@ -22,13 +22,12 @@ def change_tab():
 
 with gr.Blocks() as demo:
```
Comment (freddyaboulton) [inline demo/blocks_plug/run.py:23]: I'm not sure if the changes to Tabs are truly backwards compatible.

Changing the demo back to what it was before breaks (shows red `Error`)

image

```
index.9e34751b.js:6412 TypeError: Cannot read properties of undefined (reading 'type')
    at walk_layout (index.9e34751b.js:6385:57)
    at index.9e34751b.js:6394:45
    at Array.map (<anonymous>)
    at walk_layout (index.9e34751b.js:6394:36)
    at async Promise.all (:7860/index 1)
    at async walk_layout (index.9e34751b.js:6394:4)
    at async Promise.all (:7860/index 0)
    at async walk_layout (index.9e34751b.js:6394:4)
    at async Promise.all (:7860/index 0)
    at async walk_layout (index.9e34751b.js:6394:4)
```

For reference, I just made the following modification to this code block

```python
with gr.Blocks() as demo:
    gr.Markdown("Three demos in one!")
    with gr.Tabs(selected=1) as tabs:
        with gr.Tab("Text Identity", id=0):
            identity_demo.render()
        with gr.Tab("Text Input", id=1):
            input_demo.render()
        with gr.Tab("Text Static", id=2):
            output_demo.render()
    btn = gr.Button("Change tab")
    btn.click(inputs=None, outputs=tabs, fn=change_tab)
```

File: demo/blocks_plug/run.py (hunk 4 above)
Comment (aliabid94) [reply demo/blocks_plug/run.py]: Good catch!

File: demo/blocks_plug/run.py (hunk 5 above)
Comment (aliabid94) [reply demo/blocks_plug/run.py]: Took me forever to figure out what was going on here! Turns out Interfaces embedded in Blocks broke, finally fixed in by moving the expected_parent logic int the BlockContext.

**Pull request merged**